mne/annotations.py

```diff
@@ -517,3 +518,85 @@ def _ensure_annotation_object(obj):
     if not isinstance(obj, Annotations):
         raise ValueError('Annotations must be an instance of '
                          'mne.Annotations. Got %s.' % obj)
+
+
+@verbose
+def events_from_annotations(raw, event_id=None, regexp=None,
+                            event_id_func=None, verbose=None):
+    """Get events and event_id from an Annotations object.
+
+    Parameters
+    ----------
+    raw : instance of Raw
+        The raw data for which Annotations are defined.
+    event_id : dict | None
+        Dictionary of string keys and integer values as used in mne.Epochs
+        to map annotation descriptions to integer event codes. Only the
+        keys present will be mapped and the annotations with other descriptions
+        will be ignored. If None, all descriptions of annotations are mapped
+        and assigned arbitrary unique integer values.
+    regexp : str | None
+        Regular expression used to filter the annotations whose
+        descriptions is a match.
+    event_id_func : None | callable
+        What to do for events not found in ``event_id``. Must take one
+        argument and return an ``int``. If None, ``event_id_func`` behaves like
+        a counter that increments every time is called.
+    verbose : bool, str, int, or None
+        If not None, override default verbose level (see
+        :func:`mne.verbose` and :ref:`Logging documentation <tut_logging>`
+        for more). Defaults to self.verbose.
+
+    Returns
+    -------
+    events : ndarray, shape (n_events, 3)
+        The events.
+    event_id : dict
+        The event_id variable that can be passed to Epochs.
+    """
+    import re
+
+    if raw.annotations is None:
+        return np.empty((0, 3), dtype=int), event_id
+
+    annotations = raw.annotations
+
+    inds = raw.time_as_index(annotations.onset, use_rounding=True,
+                             origin=annotations.orig_time) + raw.first_samp
+
+    # Filter out the annotations that do not match regexp
+    if regexp is not None:
+        pat = re.compile(regexp)
```

Comment:
Branching can be reduced by using an all-matching `regex` like:
```
regexp = re.compile('.*' if regexp is None else regexp)
```
Then always check `if regexp.match(desc) is not None`